### PR TITLE
feat: refactor codes and add missing licenses

### PR DESCRIPTION
**Pull request by Xunzhuo**
+ Refactor codes with better project structure
+ Provide cobra commands and flags to make it easy to extend
+ Add missing licenses
image

> TODO: add licenses checker CI

Signed-off-by: Xunzhuo <mixdeers@gmail.com>

### CODE REVIEW

File: app/cmd/options/options.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+func NewOptions() kubernetes.Interface {
+	var err error
+	if config.Debug {
+		log.SetLevel(log.DebugLevel)
+	}
+	if config.Mode != config.ModeIstio && config.Mode != config.ModeLinkerd {
+		log.Errorf("invalid mode %q, current only support istio and linkerd", config.Mode)
+		os.Exit(1)
+	}
+	if err := ebpfs.LoadMBProgs(config.Mode, config.UseReconnect, config.Debug); err != nil {
```
Comment (kebe7jun) [inline app/cmd/options/options.go:40]: I think `LoadMBProgs` can not be in this func.

File: app/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	})
+	log.SetOutput(os.Stdout)
+	log.SetLevel(log.InfoLevel)
+	log.SetReportCaller(true)
+
+	// Get some flags from commands
+	rootCmd.PersistentFlags().StringVarP(&config.Mode, "mode", "m", config.ModeIstio, "Service mesh mode, current support istio and linkerd")
+	rootCmd.PersistentFlags().BoolVarP(&config.UseReconnect, "use-reconnect", "r", true, "Use re-connect mode as same-node acceleration")
+	rootCmd.PersistentFlags().BoolVarP(&config.Debug, "debug", "d", false, "Debug mode")
+	rootCmd.PersistentFlags().BoolVarP(&config.IsKind, "kind", "k", false, "Kubernetes in Kind mode")
```
Comment (kebe7jun) [inline app/cmd/root.go:86]: When this is not very common and the developer needs to be able to clearly understand what he or she is doing, I think we should not use short name.

**Review by kebe7jun — COMMENTED** [review]
Can we add a script to check license was be added to per file.
And run in CI.

**Comment by Xunzhuo** [comment]
> Can we add a script to check license was be added to per file.
> 
> And run in CI.

Sure, this is a todo PR, which I will separate and start this PR alone, because it is in infra area.

**Comment by Xunzhuo** [comment]
I will reach it out to update this PR later this day.

File: app/cmd/root.go (hunk 2 above)
Comment (Xunzhuo) [reply app/cmd/root.go:86]: pflags offers the full name and short name at the same time, which means if execute -h, you will see that:
```
Use eBPF to speed up your Service Mesh like crossing an Einstein-Rosen Bridge.

Usage:
  mbctl [flags]

Flags:
  -d, --debug             Debug mode
  -h, --help              help for mbctl
  -f, --ips-file string   Current node ips file name
  -k, --kind              Kubernetes in Kind mode
  -m, --mode string       Service mesh mode, current support istio and linkerd (default "istio")
  -r, --use-reconnect     Use re-connect mode as same-node acceleration (default true)
``

you can use the full name like `--mode istio` and also `-m istio`, if someone is not familiar with the flags they can use full name flag, if someone is familiar with flags, short name is better.

File: app/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+var rootCmd = &cobra.Command{
+	Use:   "mbctl",
+	Short: "Use eBPF to speed up your Service Mesh like crossing an Einstein-Rosen Bridge.",
+	Long:  `Use eBPF to speed up your Service Mesh like crossing an Einstein-Rosen Bridge.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		client := options.NewOptions()
+		w := pods.NewWatcher(client)
+
+		_ = w.Start()
+		log.Info("Pod Watcher Ready")
```
Comment (kebe7jun) [inline app/cmd/root.go:45]: this line conflicts with https://github.com/merbridge/merbridge/blob/3f10d6f82efa290d6b31ddfcfff6b6e0762c7c7c/.github/workflows/e2e.yaml#L53

File: internal/pods/watcher.go [hunk 4]
```diff
@@ -37,16 +60,56 @@ func (w *watcher) Stop() {
...
+		podHostIP := pod.Status.HostIP
+		if config.CurrentNodeIP == "" {
+			if linux.IsCurrentNodeIP(podHostIP, config.IpsFile) {
+				config.CurrentNodeIP = podHostIP
+			}
+		}
+		if podHostIP == config.CurrentNodeIP || config.IsKind {
+			_ip, _ := linux.IP2Linux(pod.Status.PodIP)
+			log.Infof("update local_pod_ips with ip: %s", pod.Status.PodIP)
+			err := config.EbpfLoadPinnedMap.Update(_ip, uint32(0), ebpf.UpdateAny)
```
Comment (kebe7jun) [inline internal/pods/watcher.go:97]: these logic should not as part of `Watcher`, meybe convert Watcher to some other name, eg: LocalIPController will be better.

File: config/vars.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+var (
+	CurrentNodeIP     string
+	Mode              string
+	IpsFile           string
+	UseReconnect      = true
+	Debug             = false
+	IsKind            = false // is Run Kubernetes in Docker
+	EbpfLoadPinnedMap *ebpf.Map
```
Comment (kebe7jun) [inline config/vars.go:33]: this doesn't seem to fit in the config,

File: app/cmd/root.go (hunk 3 above)
Comment (kebe7jun) [reply app/cmd/root.go:45]: which cause e2e test timeout.

File: .github/workflows/e2e.yaml [hunk 6]
```diff
@@ -49,8 +49,8 @@ jobs:
         kubectl exec $(kubectl get po -l app=sleep -o=jsonpath='{..metadata.name}') -c sleep -- curl -s -v helloworld:5000/hello
     - name: install merbridge
       run: |
-        nohup go run -exec sudo ./cmd/mbctl/main.go --kind -m istio -d > mbctl.log &
-        while true; do [ "$(cat mbctl.log | grep 'pod watcher ready')" = "" ] || break && (echo waiting for mbctl watcher ready; sleep 3); done
+        nohup go run -exec sudo ./app/main.go -k true -m istio -d true > mbctl.log &
```
Comment (kebe7jun) [inline .github/workflows/e2e.yaml:52]: `-k true` means with `-k` flag, and a `"true"` args...
```suggestion
        nohup go run -exec sudo ./app/main.go -k -m istio -d > mbctl.log &
```

**Review by kebe7jun — APPROVED** [review]

**Comment by yankay** [comment]
Should we use ”addlicense“ tool to make it automately

**Pull request merged**